lib/Sema/MiscDiagnostics.cpp

```diff
@@ -4987,6 +4987,168 @@ static void diagUnqualifiedAccessToMethodNamedSelf(const Expr *E,
   const_cast<Expr *>(E)->walk(Walker);
 }
 
+static void
+diagnoseDictionaryLiteralDuplicatedKeyEntries(const Expr *E,
+                                              const DeclContext *DC) {
+  class DiagnoseWalker : public ASTWalker {
+    ASTContext &Ctx;
+
+  private:
+    llvm::Optional<std::string> getKeyStringValue(const Expr *keyExpr) {
+      switch (keyExpr->getKind()) {
+      case ExprKind::StringLiteral:
+      case ExprKind::FloatLiteral:
+      case ExprKind::IntegerLiteral:
+      case ExprKind::BooleanLiteral:
+      case ExprKind::NilLiteral: {
+        std::string out;
+        llvm::raw_string_ostream OS(out);
+        keyExpr->printConstExprValue(&OS, /*additionalCheck=*/nullptr);
+        return out;
+      }
+      default:
+        break;
+      }
+      return None;
+    }
+
+    std::string getKeyLiteralDescription(const Expr *keyExpr) {
```

Comment:
There are a couple of places in the Sema where this exact operation is needed i.e. for binding printing and for diagnostics (search for `ExprKind::IntegerLiteral`). Looks like it's time to move it to a hear and start using it from all the places where this is needed...